app/src/main/java/org/mozilla/focus/searchwidget/SearchWidgetNavigation.kt

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.searchwidget
+
+import android.app.Activity
+import android.os.Bundle
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.feature.search.widget.BaseVoiceSearchActivity
+import mozilla.telemetry.glean.private.NoExtras
+import org.mozilla.focus.GleanMetrics.SearchWidget
+import org.mozilla.focus.activity.IntentReceiverActivity
+import org.mozilla.focus.ext.components
+import org.mozilla.focus.ext.settings
+import org.mozilla.focus.perf.Performance
+import org.mozilla.focus.state.AppAction
+import org.mozilla.focus.utils.SearchUtils
+
+object SearchWidgetNavigation {
+
+    /**
+     * Method used to handle all the navigation from Search widget ,app icon ,lockscreen and
+     * custom tab.
+     */
+    fun handleAppNavigation(
+        bundle: Bundle?,
+        activity: Activity,
+        navigator: () -> Unit,
+        splashScreenPreDrawListener: () -> Unit,
+        unlock: () -> Unit,
+    ) {
+        val voiceSearchText = bundle?.getString(BaseVoiceSearchActivity.SPEECH_PROCESSING)
```

Comment:
nit: It seems like we have four cases. Can they be better grouped by using spaces maybe together with a small comment about what scenario each case evaluates? This would hopefully help with helping other down the line and prevent regressions. 
Giving the importance of it unit tests are more than welcomed.

Reply:
Done